### PR TITLE
Feature/collison monitoring

### PR DESCRIPTION
**Pull request by r2avula**
>  **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- Feature implementing onboard collision monitoring using lidar scan and depth camera point cloud. 

- Also replaced teleop_twist_keyboard packaage with a new twist_keyboard node in waywiser_teleop.

> **What is the current behavior?** (You can also link to an open issue here)

- Collision monitoring does not exist. 

- teleop_twist_keyboard node only publishes twist messages when a keyboard key is pressed. This does not seem to update the polygons topic in nav2 collision monitoring and allowing the vehicle to collide with obstacles. Also, multiple key presses are not recognized with teleop_twist_keyboard. Consequetly, it is only possible to set either linear or angular speed in a twist message. 

> **What is the new behavior (if this is a feature change)?**

- The input twist commands will be filtered based on obstacle presence within a defined zone (circle/polygon) around the vehicle. It can be configured to approach, slow down, and stop based on distance and time to collision. 
- A new twist_keyboard node is implemented in waywiser_teleop that continuously publishes twist messages at a set rate. Also multiple key presses are recognized, allowing to set both linear and angular speeds simultaneously. 

> **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No.

### CODE REVIEW

File: waywiser_command_control/config/command_control.yaml [hunk 1]
```diff
@@ -13,3 +13,61 @@ onboard_twist_mux:
         topic: waywise_vel
         timeout: 0.5
         priority: 60
+
+collision_monitor:
```
Comment (marvind) [inline waywiser_command_control/config/command_control.yaml]: If this is based on an existing external file, I would like to see a comment here referencing it. Is this tuned to our setup in any way? A link to documentation, e.g., on how to adapt / fine tune this to different vehicles would be great.

File: waywiser_command_control/launch/command_control.launch.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ def generate_launch_description():
                 'use_sim_time': LaunchConfiguration('use_sim_time'),
             },
         ],
-        remappings={('/cmd_vel_in', '/onboard_mux_vel')},
+        remappings={('/cmd_vel_in', '/cmd_vel_safe')},
```
Comment (marvind) [inline waywiser_command_control/launch/command_control.launch.py]: It would be great to have a node graph as part of the documentation of waywiser_command_control. It is IMO quite hard to follow how the nodes are connected and which node does what, this should also be updated in the package.xml. Is it now mandatory to run the whole set of nodes or can you, e.g., run emergency stop only?

I would prefer naming cmd_vel_safe s.th. like cmd_vel_collisionmon as cmd_vel_safe suggests it is safe already (why should it go to emergency stop then?).

File: waywiser_command_control/launch/command_control.launch.py [hunk 3]
```diff
@@ -48,6 +48,33 @@ def generate_launch_description():
...
+        output='screen',
+        emulate_tty=True,  # https://github.com/ros2/launch/issues/188
+        parameters=[
+            LaunchConfiguration('command_control_config'),
+            {'use_sim_time': LaunchConfiguration('use_sim_time')},
+        ],
+        remappings={('/bond', '/bond_collision_monitor')},
+    )
+
+    lifecycle_manager_collision_monitor = Node(
```
Comment (marvind) [inline waywiser_command_control/launch/command_control.launch.py]: I am not sure why this is needed? Is it always needed for collision_monitor?

File: waywiser_command_control/launch/command_control.launch.py [hunk 4]
```diff
@@ -58,5 +85,7 @@ def generate_launch_description():
     # start nodes
     ld.add_action(emergency_stop_monitor_node)
     ld.add_action(onboard_twist_mux_node)
+    ld.add_action(collision_monitor)
+    ld.add_action(lifecycle_manager_collision_monitor)
```
Comment (marvind) [inline waywiser_command_control/launch/command_control.launch.py]: *_node for consistency?

File: waywiser_command_control/package.xml [hunk 5]
```diff
@@ -21,6 +21,8 @@
   <depend>std_msgs</depend>
```
Comment (marvind) [inline waywiser_command_control/package.xml]: I am wondering whether waywiser_command_control is a good name, isn't it more like waywiser_cmd_vel_safety? To my understanding it is the vehicle side of processing cmd_vel safely.

File: waywiser_hwbringup/package.xml [hunk 6]
```diff
@@ -8,16 +8,26 @@
...
+  <build_depend>tf2_sensor_msgs</build_depend>
+  <build_depend>geometry_msgs</build_depend>
+
   <depend>joint_state_publisher</depend>
   <depend>robot_state_publisher</depend>
   <depend>rplidar_ros</depend>
+  <depend>librealsense2-utils</depend>
+  <depend>librealsense2-dev</depend>
+  <depend>realsense2_camera</depend>
+  <depend>realsense2_camera_msgs</depend>
```
Comment (marvind) [inline waywiser_hwbringup/package.xml]: General thought on specific hardware: sooner or later we will end up with a lot of dependencies here. Is it possible to have optional dependencies? I am thinking that you can run the default setup (no lidar nor camera) without them. The specific hw nodes only need to be available when you acutally want to enable them.

File: waywiser_hwbringup/include/point_cloud_transformer.hpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#include "sensor_msgs/msg/point_cloud2.hpp"
+#include "tf2_ros/buffer.h"
+#include "tf2_ros/transform_listener.h"
+#include "tf2_sensor_msgs/tf2_sensor_msgs.hpp"
+#include "tf2/exceptions.h"
+#include "geometry_msgs/msg/transform_stamped.hpp"
+
+namespace waywiser_hwbringup
+{
+class PointCloudTransformer : public rclcpp::Node
```
Comment (marvind) [inline waywiser_hwbringup/include/point_cloud_transformer.hpp:20]: I would like to have a short comment about what it does and why it is needed somewhere here in the actual node, not only launch file. Is this an actual "fix" and realsense specific or does it need to be applied to any depth camera (looks like the latter but not entirely sure)?
Are there any benefits distributing this node to three files instead of one (max two) considering its small size?

File: waywiser_teleop/py_scripts/twist_keyboard.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
```
Comment (marvind) [inline waywiser_teleop/py_scripts/twist_keyboard.py:1]: That this is python makes me suspicious. 😄 Is this based on something existing that we should to reference?

**Review by marvind — CHANGES_REQUESTED** [review]
I did a dry run now, going through the code without testing. It looks great so far. 😎
I have minor comments concerning clarity and adding a bit of documentation.

@ariamirzai, do you have time to build and test the functionality?

File: waywiser_command_control/config/command_control.yaml (hunk 1 above)
Comment (r2avula) [reply waywiser_command_control/config/command_control.yaml]: The polygon shapes are tuned for the RC car. Unfortunately I didn't find any documentation for fine tuning the polygons for humble release. There exists configuration guide from nav2 but it refers to a later release and contains more parameters than the humble release. I believe that "https://github.com/ros-planning/navigation2/tree/humble/nav2_collision_monitor" is the most accurate link for documentation. I added it in the config file and in the package.xml.

File: waywiser_command_control/launch/command_control.launch.py (hunk 4 above)
Comment (r2avula) [reply waywiser_command_control/launch/command_control.launch.py]: updated.

File: waywiser_command_control/launch/command_control.launch.py (hunk 3 above)
Comment (r2avula) [reply waywiser_command_control/launch/command_control.launch.py]: Yes, nav2 collision_monitor is a managed node and always requires a life cycle manager. (https://design.ros2.org/articles/node_lifecycle.html)

File: waywiser_command_control/launch/command_control.launch.py (hunk 2 above)
Comment (r2avula) [reply waywiser_command_control/launch/command_control.launch.py]: Updated the readme file and package.xml. It is not mandatory to run all nodes in command_control launch file. I added a launch argument that can be set to choose whether to include collision monitoring. Nonetheless, launching all the nodes along with the collision monitor would still work even if the RC car does not have any lidar or depth camera. Renamed cmd_vel_safe to collision_monitor_vel.

File: waywiser_command_control/package.xml (hunk 5 above)
Comment (r2avula) [reply waywiser_command_control/package.xml]: Package renamed to waywiser_twist_safety.

File: waywiser_teleop/py_scripts/twist_keyboard.py (hunk 8 above)
Comment (r2avula) [reply waywiser_teleop/py_scripts/twist_keyboard.py:1]: It was a bit surprising that I couldn't find any standard cpp library to record multiple key presses, but it works like a charm with pygame python module. It wasn't based on any existing work.

File: waywiser_hwbringup/include/point_cloud_transformer.hpp (hunk 7 above)
Comment (r2avula) [reply waywiser_hwbringup/include/point_cloud_transformer.hpp:20]: It is actually a fix for correct visualization of point clouds in rviz2. To represent the points using their depth value for their color intensity, we would have to choose "z" axis instead of "x" axis (as should be done if the data is in its ros2 base frame). This seems to be the case with the realsense depth camera where the reference frame id of point cloud is not its base frame ("camera_link"). I am not sure if it is the case with other depth cameras. 

This fix is not otherwise required since this conversion is done where the point cloud is used. For example, when using collision monitor, it is done here: https://github.com/ros-planning/navigation2/blob/b489f5b79baea6c9f6b91ecb23c5a2725317c140/nav2_collision_monitor/src/pointcloud.cpp#L112

File: waywiser_hwbringup/package.xml (hunk 6 above)
Comment (r2avula) [reply waywiser_hwbringup/package.xml]: It is a good point. I remove these hardware related dependencies from package.xml and added them in readme file.

**Comment by marvind** [comment]
Thanks for the explanations and adjustments. Great job! 😊
Good to merge once @ariamirzai gives his OK. 🙂

**Review by ariamirzai — COMMENTED** [review]
Couldn't test collision detection on the real vehicle, but otherwise it looks good!

**Review by marvind — APPROVED** [review]

**Review by ariamirzai — APPROVED** [review]
